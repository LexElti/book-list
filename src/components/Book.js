import React from 'react'

export default ({book}) => {
  return (
    <div className="col-lg-3 col-md-4 col-sm-6 mb-4">
      <div className="card">
        {book.thumbnail  
          ? <img src={book.thumbnail} className="card-img-top" />
          : <div className="card-img-top"></div>
        }
        <div className="card-body">
          <h5 className="card-title">{book.title}</h5>
          <p className="card-text">Authors: {book.authors.join(', ')}</p>
          <p className="card-text">{book.description}</p>
          <a 
            href={book.previewLink}
            className="btn btn-primary"
            target="_blank"
            rel="noopener noreferrer"
          >Preview</a>
        </div>
      </div>
    </div>
  )
}
