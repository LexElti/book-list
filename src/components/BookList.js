import React from 'react'
import {connect} from 'react-redux'
import Book from './Book'
import Loader from './Loader'

export const BookList = ({itemCount, books, page, loader}) => {
  if (loader) {
    return <Loader />
  }
  
  const fistIndex = page * itemCount - itemCount
  const lastIndex = page * itemCount - 1
  const currentBooks = books.filter((_, index) => index >= fistIndex && index <= lastIndex)

  return currentBooks.map(book => <Book book={book} key={book.id} />)
}

const mapStateToProps = state => ({
  itemCount: state.books.itemCount,
  books: state.books.books,
  page: state.books.page,
  loader: state.app.loading
})

export default connect(mapStateToProps, null)(BookList)
