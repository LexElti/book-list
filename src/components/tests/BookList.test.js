import React from 'react'
import {BookList} from '../BookList'
import Loader from '../Loader'
import Book from '../Book'

describe('<BookList />', () => {
  const props = {
    itemCount: 8,
    books: [],
    page: 1,
    loader: false
  }

  describe('<BookList /> initial', () => {
    const wrapper = shallow(<BookList {...props} />)

    it('renders properly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should have components', () => {
      expect(wrapper.find(Book)).toHaveLength(0)
    })

    it('should not render Loader', () => {
      expect(wrapper.find(Loader)).toHaveLength(0)
    })
  })

  describe('<BookList /> with loader', () => {
    const nextProps = {
      ...props,
      loader: true,
    }

    const wrapper = shallow(<BookList {...nextProps} />)

    it('renders properly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should have components', () => {
      expect(wrapper.find(Book)).toHaveLength(0)
    })

    it('should render Loader', () => {
      expect(wrapper.find(Loader)).toHaveLength(1)
    })
  })

  describe('<BookList /> with books', () => {
    const books = Array(20).fill('')
    const nextProps = {
      ...props,
      books
    }

    const wrapper = shallow(<BookList {...nextProps} />)

    it('renders properly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should have components', () => {
      expect(wrapper.find(Book)).toHaveLength(8)
    })

    it('should not render Loader', () => {
      expect(wrapper.find(Loader)).toHaveLength(0)
    })
  })
})
