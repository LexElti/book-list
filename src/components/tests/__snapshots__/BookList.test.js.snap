// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookList /> <BookList /> initial renders properly 1`] = `ShallowWrapper {}`;

exports[`<BookList /> <BookList /> with books renders properly 1`] = `ShallowWrapper {}`;

exports[`<BookList /> <BookList /> with loader renders properly 1`] = `ShallowWrapper {}`;
